apps/application-system/api/src/app/modules/application/tools/applicationAccess.service.ts

```diff
@@ -22,4 +46,77 @@ export class ApplicationAccessService {
 
     return existingApplication as Application
   }
+
+  async findOneByIdAndDelegations(id: string, auth: Auth) {
+    const actorDelegations = await this.delegationsApiWithAuth(
+      auth,
+    ).actorDelegationsControllerFindAll({
+      direction: ActorDelegationsControllerFindAllDirectionEnum.Incoming,
+    })
```

Comment:
We need to add guard logic to most application controller endpoints to make sure delegations can't access or update application which don't allow the current delegation (403 Forbidden).

    E.g. so I can't sign into my child and apply for parental leave.

I would like to get rid of the new `/delegated/:id` endpoint and move this logic (or similar) into `ApplicationAccessService#findOneByIdAndNationalId` so the applicationApplication query returns the correct problem.

Like mentioned elsewhere, we don't need to depend on Auth Public API here, and we shouldn't use the deprecated `delegation.fromName` and `delegation.toName`.

This logic can be much simpler. Just check if `user.nationalId` or `actor?.nationalId` is in the `application.applicantActors` array, then throw our fancy problem with the `application.applicant` nationalId.

Reply:
Thank you for the feedback! I think these are really good ideas and a good direction, making this as simple as it can be is the goal 😄  
One thing I question is 

_Just check if user.nationalId or actor?.nationalId is in the application.applicantActors array, then throw our fancy problem with the application.applicant nationalId._

Since more than one person can be actors for the same subject. If one actor or the subject itself creates the application then the second actor would not be assigned in the `applicantActors` array and the error would not be thrown. 
How can we check if the user has the `application.applicant `in their delegation, and therefore should get the fancy problem if we don't check if the applicant exists in their delegations in the `applicationAccess.service`

For all other use cases your suggestions are spot on!

Reply:
Great point. IMHO this is an edge case (second actor joining an existing application) we should either not worry about, or solve in a different way.

This only happens if someone opens a direct link to an application they have a delegation for but have never opened before. Typically, the user would sign into My Pages, switch to the delegation, find the application there to open it, in which case they'd drop right in.

I think this is rare enough to warrant simplifying the code and skipping the extra Auth Public API integration. 

A different way would be to have a "generic" delegation switcher on some application system "Access Denied" page. Google Drive does something similar: "You do not have access. You are signed in as Eirikur Nilsson. Switch accounts." We can show a delegation dropdown right there, or a simple link which sends the user to the IDP delegation prompt:

<img width="580" alt="image" src="https://user-images.githubusercontent.com/115094/160656830-de26afa1-2996-4b78-b107-554bcc907da3.png">

Reply:
This is a good point! Let's go with that and try to handle it else where. Project simplification STARTS NOW 🦾